Adds gerrit CLI script and some basic functions

polynomial opened the pull request:
Adds gerrit CLI script and some basic functions like:
- add reviewers to a review
- comment on a review

Need feedback on path for z-gerrit/z-review, specifically:
- do we want to abstract this further so z-review knows nothing
  about gerrit and only uses some sort of lib/bash/review.functions
  that talks to gerrit
- do we want to support doing stuff outside the repo by ID only?
- don't think we need it now, but having OWNERS support would be
  awesome, discuss
- should we add functionality that wraps the add/commit process
  too? like git flow 'can' do (protects engineers who can't seem
  to ever remember to rebase first and always ask why they are
  stuck in recursive merge land)

Commit:
Adds gerrit CLI script and some basic functions like:
 * add reviewers to a review
 * comment on a review

Need feedback on path for z-gerrit/z-review, specifically:
 * do we want to abstract this further so z-review knows nothing
   about gerrit and only uses some sort of lib/bash/review.functions
   that talks to gerrit
 * do we want to support doing stuff outside the repo by ID only?
 * don't think we need it now, but having OWNERS support would be
   awesome, discuss
 * should we add functionality that wraps the add/commit process
   too? like git flow 'can' do (protects engineers who can't seem
   to ever remember to rebase first and always ask why they are
   stuck in recursive merge land)
 * Moved logic out of z- scripts into functions where possible
 * Added help stdout/stderr functions
 * Improved error checking inside of env so there is less boilerplate
   in scripts

polynomial commented:
@mbbx6spp sorry for being so slow with this, but I had to make sure it was absolutely horrible.

mbbx6spp commented on bin/z-gerrit:
```diff
@@ -0,0 +1,53 @@
+
+. $Z_LIB/gerrit.functions
+
+usage="Usage: $(basename $0) [-d] [gerrit_command]
```
This variable isn't declared with appropriate flags (e.g. read-only). I actually prefer putting usage in a function that echos to stderr or stdout depending on when and how it is called (taking a parameter for the stream to output to).

mbbx6spp commented on bin/z-gerrit:
```diff
@@ -0,0 +1,53 @@
+
+set -e
+
+if [ -n "$Z_LIB" ]; then
```
This should be in a function that can be shared by all scripts and a function to check `Z_LIB`, `Z_HOME`, etc. to validate the environment is setup properly and then the file just sourced. I think? Thoughts?

mbbx6spp commented on bin/z-gerrit:
```diff
@@ -0,0 +1,53 @@
+
+while [ $# -gt 0 ]; do
+  case "$1" in
+    -h)
```
Also add long version to case, e.g.:

``` bash
case "${1}" in
  -h|--help) ... ;;
  -d|--debug) ... ;;
  *) ... ;;
esac
```

mbbx6spp commented on bin/z-gerrit:
```diff
@@ -0,0 +1,53 @@
+  exit 1
+fi
+
+gerrit_prefix="ssh -To BatchMode=yes ${gerrit_host} gerrit"
```
Should be a function that accepts arguments and arguments are checked for sanity instead of an undeclared variable. :)

mbbx6spp commented on bin/z-gerrit:
```diff
@@ -0,0 +1,53 @@
+
+gerrit_prefix="ssh -To BatchMode=yes ${gerrit_host} gerrit"
+
+if [ -n "$1" ]; then
```
This kind of logic can be wrapped in inside said function (in directly above comment).

mbbx6spp commented on bin/z-review-comment:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
+
```
Generally speaking code needs to be put in more functions, some of which could be reusable and moved to a library file instead and exported. I think. Thoughts?

mbbx6spp commented on lib/bash/bootstrap.sh:
```diff
@@ -0,0 +1,33 @@
+    export -r Z_HOME=$(dirname $(dirname $(dirname $0)))
+    export -r Z_BIN=${Z_HOME}/bin
+    export -r Z_LIB=${Z_HOME}/lib/bash
+    alias cd-="cd $Z_HOME"
```
I think this alias is dangerous.

mbbx6spp replied:
Nevermind I misread and didn't see the dash. Phew.

polynomial commented:
Ok, I think I did all the feedback, again in the most disappointing way possible.

mbbx6spp commented on lib/bash/bootstrap.sh:
```diff
@@ -0,0 +1,33 @@
+  if [ -z "$Z_HOME" ]; then
+    export -r Z_HOME=$(dirname $(dirname $(dirname $0)))
+    export -r Z_BIN=${Z_HOME}/bin
+    export -r Z_LIB=${Z_HOME}/lib/bash
```
I think `Z_HOME`, `Z_BIN`, `Z_LIB` should go into env.sh.

mbbx6spp replied:
Also Z_LIB might be better off defined as `${Z_HOME}/lib` instead of `${Z_HOME}/lib/bash` and then we can have a `${Z_BASH_LIB}` for the above value.

mbbx6spp commented on lib/bash/bootstrap.sh:
```diff
@@ -0,0 +1,33 @@
+    export -r Z_LIB=${Z_HOME}/lib/bash
+    alias cd-="cd $Z_HOME"
+    alias cdb="cd $Z_BIN"
+    alias cdl="cd $Z_LIB"
```
Same with these aliases and the path update below. What do you think?

mbbx6spp commented:
Minor comments now which are more open questions. I can be persuaded.

mbbx6spp commented:
One last comment, please squash when making next changes, I think we'll be done :)

polynomial commented:
Squashed and cleaned up LIB. Re: other comments let me try the persuading (: I see env.sh as the stuff you need to do work inside the scripts. bootstrap.sh is the stuff you need to call these scripts from a shell. So stuff that will 'dirty' the users environment, like aliases, Z_HOME/Z_BASH_LIB will go in there. Anything that the scripts themselves need will go in env. This provides a nice distinction which wasn't clear to others working on this stuff before.

mbbx6spp commented:
Then we need a third option. Right now I am not comfortable `bootstrap` will be written in an idempotent way. To me it has a fundamentally different purpose which is, when a new developer starts using this (or any other) repository it sets up their dev environment. So we need a third location. Perhaps it is simply `bootstrap.sh` for just simple declarations which is itself sourced by `bootstrap` executable script, but they must be separate.

mbbx6spp commented:
Ok after discussing, it seems I am on drugs when I made that comment.
